### PR TITLE
[Improve][Connector][PulsarSource]Improve pulsar deserialization

### PR DESCRIPTION
**Pull request by lightzhao**
## Purpose of this pull request
Improve pulsar deserialization
1.Add format and field_delimiter configuration.
2.Rich deserialization methods.

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java [hunk 1]
```diff
@@ -226,10 +234,37 @@ private void setPartitionDiscoverer(Config config) {
...
+                if (config.hasPath(FIELD_DELIMITER.key())) {
+                    delimiter = config.getString(FIELD_DELIMITER.key());
+                }
+                deserialization = (DeserializationSchema<T>) TextDeserializationSchema.builder()
+                        .seaTunnelRowType(rowType)
+                        .delimiter(delimiter)
+                        .build();
+            } else {
+                // TODO: use format SPI
+                throw new SeaTunnelJsonFormatException(CommonErrorCode.UNSUPPORTED_DATA_TYPE,
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java]: Can't throw UNSUPPORTED_DATA_TYPE, CommonErrorCode.UNSUPPORTED_OPERATION maybe better.

**Comment by Hisoka-X** [comment]
Please add change to change log in pulsar doc

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java (hunk 1 above)
Comment (lightzhao) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/source/PulsarSource.java]: done.

**Comment by lightzhao** [comment]
> Please add change to change log in pulsar doc

done.

File: docs/en/connector-v2/source/pulsar.md [hunk 2]
```diff
@@ -152,3 +163,4 @@ source {
 
 ### 2.3.0-beta 2022-10-20
 - Add Pulsar Source Connector
+- Improve pulsar deserialization([3990](https://github.com/apache/incubator-seatunnel/pull/3990))
```
Comment (Hisoka-X) [inline docs/en/connector-v2/source/pulsar.md:170]: It should in next version
```suggestion
### Next Version
- Improve pulsar deserialization([3990](https://github.com/apache/incubator-seatunnel/pull/3990))
```

**Comment by Hisoka-X** [comment]
Please add schema / format config into e2e test case.

**Comment by lightzhao** [comment]
> Please add schema / format config into e2e test case.

@Hisoka-X  add complete.

**Comment by Hisoka-X** [comment]
Hi, please fix code style.

**Comment by lightzhao** [comment]
> Hi, please fix code style.

done.

**Comment by lightzhao** [comment]
@Hisoka-X @ashulin @TyrantLucifer PTAL.

**Comment by lightzhao** [comment]
@EricJoy2048 @Hisoka-X @ashulin @TyrantLucifer PTAL.

**Review by TyrantLucifer — CHANGES_REQUESTED** [review]
Please fix the ci error.

**Comment by lightzhao** [comment]
> Please fix the ci error.
I don't know why these two ci errors occur occasionally. My local compilation is all right.
image

**Comment by lightzhao** [comment]
@ashulin PTAL.

**Review by TyrantLucifer — CHANGES_REQUESTED** [review]
Please ensure that your pull request's CI/CD process passes. Thank you.

**Comment by lightzhao** [comment]
> Please ensure that your pull request's CI/CD process passes. Thank you.
I have merged the latest dev branch, and the following link is always reported as abnormal, but the access to this link is normal. Do you know what the problem is?
image

**Comment by Hisoka-X** [comment]
Please fix ci

**Comment by lightzhao** [comment]
> Please fix ci

Cannot perform ci check after update.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        <groupId>org.apache.seatunnel</groupId>
+        <artifactId>seatunnel-connector-v2-e2e</artifactId>
+        <version>${revision}</version>
+    </parent>
+
+    <artifactId>connector-pulsar-e2e</artifactId>
+
+    <properties>
+        <maven.compiler.source>11</maven.compiler.source>
+        <maven.compiler.target>11</maven.compiler.target>
```
Comment (ashulin) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml]:
```suggestion
```
The root pom.xml will manage these properties uniformly.

Comment (lightzhao) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml]: ok, updated.

**Comment by lightzhao** [comment]
please approve ci check.

**Comment by lightzhao** [comment]
@Hisoka-X please approve ci check.

**Comment by lightzhao** [comment]
@Hisoka-X @TyrantLucifer @ashulin all checks passed , PTAL.

**Comment by lightzhao** [comment]
Hello everyone, who has time to help review this pr. thanks.

**Comment by lightzhao** [comment]
@TyrantLucifer @Hisoka-X @ashulin  PTAL , thanks.

**Comment by lightzhao** [comment]
Hello everyone, who has time to help review this pr. thanks. @EricJoy2048 @TyrantLucifer @Hisoka-X @ashulin

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.seatunnel</groupId>
+            <artifactId>connector-assert</artifactId>
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.testcontainers</groupId>
```
Comment (TyrantLucifer) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml]: Why add this?

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            <artifactId>testcontainers</artifactId>
+            <version>1.17.6</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
+            <scope>compile</scope>
```
Comment (TyrantLucifer) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml]: Why add this?

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <version>3.23.1</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>pulsar</artifactId>
+            <version>1.17.6</version>
```
Comment (TyrantLucifer) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml]: Use properties

Comment (TyrantLucifer) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml]: Use properties

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>jul-to-slf4j</artifactId>
+            <version>${slf4j.version}</version>
```
Comment (TyrantLucifer) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml:99]: Why add this?

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml (hunk 4 above)
Comment (lightzhao) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml]: deleted.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml (hunk 5 above)
Comment (lightzhao) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml]: deleted.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml (hunk 6 above)
Comment (lightzhao) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml]: done.

Comment (lightzhao) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml]: done.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml (hunk 7 above)
Comment (lightzhao) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml:99]: This class is needed when the test is running, and this log class is missing when ci.

**Comment by lightzhao** [comment]
@EricJoy2048 @TyrantLucifer @Hisoka-X @ashulin Please approve CI check,thanks.

**Comment by lightzhao** [comment]
Check did not find a failure, why is it prompted 'Some checks were not successful', can you retry the check.

**Comment by Hisoka-X** [comment]
> Check did not find a failure, why is it prompted 'Some checks were not successful', can you retry the check.

Some check timout and be canceled. Never mind, ci already pass.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml (hunk 7 above)
Comment (TyrantLucifer) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml:99]: cc @hailin0

Comment (lightzhao) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml:99]: @hailin0 PTAL.

Comment (lightzhao) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml:99]: @hailin0 PTAL. thanks.
This pr time is quite long, there are other pr waiting for it.

**Comment by lightzhao** [comment]
@EricJoy2048 @TyrantLucifer @Hisoka-X @ashulin PTAL,thanks.

**Comment by lightzhao** [comment]
hi，all，Who has time to help review this pr, it takes a long time, I can add e2e test to #4382 PulsarSink after completion.

**Comment by lightzhao** [comment]
@EricJoy2048 @TyrantLucifer @Hisoka-X @ashulin PTAL, thanks.

**Comment by lightzhao** [comment]
Is there a problem with e2e? 
image

**Comment by lightzhao** [comment]
please approve ci. thanks.

**Comment by lightzhao** [comment]
@Hisoka-X @TyrantLucifer @hailin0 please approve ci. thanks.

**Comment by Hisoka-X** [comment]
> @Hisoka-X @TyrantLucifer @hailin0 please approve ci. thanks.

Sorry for late response, done

**Comment by lightzhao** [comment]
@Hisoka-X @TyrantLucifer @ashulin @hailin0 @EricJoy2048 PTAL, thanks.

**Comment by lightzhao** [comment]
> Please ensure that your pull request's CI/CD process passes. Thank you.

done.

**Comment by lightzhao** [comment]
@Hisoka-X @TyrantLucifer @ashulin @hailin0 @EricJoy2048 PTAL, thanks.

**Comment by lightzhao** [comment]
Hi all ,Who has time to review , it's been a long time, #4382 pulsar sink is waiting for this pr.

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/config/SourceProperties.java [hunk 8]
```diff
@@ -167,8 +174,20 @@ public class SourceProperties {
                     .longType()
                     .noDefaultValue()
                     .withDescription("Stop from the specified epoch timestamp (in milliseconds)");
+    public static final Option<String> FORMAT =
+            Options.key("format")
+                    .stringType()
+                    .noDefaultValue()
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/config/SourceProperties.java:180]: No defaule value? Or the default value is `json`? I suggest define the default value in option.  `public static final String DEFAULT_FORMAT = "json";` is not a good way.

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/config/SourceProperties.java [hunk 9]
```diff
@@ -167,8 +174,20 @@ public class SourceProperties {
...
+                    .withDescription(
+                            "Data format. The default format is json. Optional text format. The default field separator is \", \". "
+                                    + "If you customize the delimiter, add the \"field_delimiter\" option.");
+
+    public static final Option<String> FIELD_DELIMITER =
+            Options.key("field_delimiter")
+                    .stringType()
+                    .noDefaultValue()
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/config/SourceProperties.java:188]: same as above.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/resources/pulsarsource_text_to_console.conf [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    # The default field delimiter is ","
+    field_delimiter = ","
+  }
+}
+
+transform {
+}
+
+sink {
+  Console {}
```
Comment (EricJoy2048) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/src/test/resources/pulsarsource_text_to_console.conf:50]: I think you need check the data read from the source. You can add a LocalFile Sink Or Pulsar Sink and then read the data from the target file(or topic when you use Pulsar Sink). Check every row and column.

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/config/SourceProperties.java (hunk 8 above)
Comment (hailin0) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/config/SourceProperties.java:180]: @lightzhao

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml (hunk 7 above)
Comment (hailin0) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-pulsar-e2e/pom.xml:99]: LGTM

File: seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/config/SourceProperties.java (hunk 8 above)
Comment (lightzhao) [reply seatunnel-connectors-v2/connector-pulsar/src/main/java/org/apache/seatunnel/connectors/seatunnel/pulsar/config/SourceProperties.java:180]: I've been a little busy recently, I'll update it in the next few days.

**Comment by lightzhao** [comment]
#4111 This ability has been merged, I will close this pr.

**Pull request closed without merging**